presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,274 @@
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Ordering;
+import com.google.common.collect.Sets;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.SystemSessionProperties.getJoinDistributionType;
+import static com.facebook.presto.SystemSessionProperties.isJoinReorderingEnabled;
+import static com.facebook.presto.sql.ExpressionUtils.extractConjuncts;
+import static com.facebook.presto.sql.planner.optimizations.InnerJoinPredicateUtils.sortPredicatesForInnerJoin;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.EQUAL;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class ReorderJoins
+        implements Rule
+{
+    private final CostComparator costComparator;
+
+    public ReorderJoins(CostComparator costComparator)
+    {
+        this.costComparator = costComparator;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof JoinGraphNode) || !isJoinReorderingEnabled(session)) {
+            return Optional.empty();
+        }
+
+        JoinGraphNode joinGraph = (JoinGraphNode) node;
+        JoinNode joinNode = new JoinEnumerator(idAllocator, symbolAllocator, session, lookup).chooseJoinOrder(joinGraph);
+        return Optional.of(joinNode);
+    }
+
+    @VisibleForTesting
+    public class JoinEnumerator
+    {
+        private final Map<String, JoinNode> memo = new HashMap<>();
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+        private final Session session;
+        private final Lookup lookup;
+
+        public JoinEnumerator(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session, Lookup lookup)
+        {
+            this.idAllocator = idAllocator;
+            this.symbolAllocator = symbolAllocator;
+            this.session = session;
+            this.lookup = lookup;
+        }
+
+        private JoinNode chooseJoinOrder(JoinGraphNode joinGraph)
+        {
+            if (memo.containsKey(joinGraph.getKey())) {
```

Comment:
You mustn't `computeIfAbsent` because computation is non side-effect-free computation that actually _modifies_ memo in recursive calls. That would blow `computeIfAbsent`.

What you _can_ is optimizing for existence:
```
join = memo.get(...);
if (join == null) {
    join = ...
    memo.put(..., join)
}
return join;
```